cloudpickle/cloudpickle.py

```diff
@@ -208,6 +238,9 @@ def _lookup_module_and_qualname(obj, name=None):
     if module_name == "__main__":
         return None
 
+    if _is_explicitly_serialized_module(module_name):
```

Comment:
I think It would be clearer if this check was made at a higher point in the cloudpickle logic.

As a first step, I think we could put this check in `is_importable`. This will require a small additional adaptation for `TypeVar` pickling. This will also change the behavior of pickling `modules` see the comment of the overall review I made.

This situation is still not ideal: Indeed, a module registered for pickling by value is a priori importable - it is just pickled by value. So the code structure/naming after the first step I explained above remains a bit confusing. An additional refactoring step would be to instead change checks in `_function_reduce` and other concerned functions from:
```python
        if _is_importable(obj):
```
to 
```python
        if _is_importable(obj) and is_registered_pickle_by_value(obj.__module__):
```
Although I'm being careless with my `obj.__module__` and should use `whichmodule` - on the other hand, It would be better to not call `whichmodule` twice since the utility may be costly.

@ogrisel this is not a nitpick IMO, so tagging you in case you have an opinion.

Reply:
A solution which gives both worlds might be to call `whichmodule` initially, and pass the result in to both functions. Possible as optional kwargs which could be checked against a sentinel value (Ellipsis perhaps as None is a valid output) to see if its been set and if not, call `whichmodule` themselves. Removes complicating the `_is_importable` signature but adds sentinel checking complexity in, which might not be wanted.

Reply:
> Removes complicating the _is_importable signature but adds sentinel checking complexity in, which might not be wanted

I think this may be prone to duplication of work being overlooked when making code changes at further point in time since passing `None` for the `module` argument will be silently handled by `cloudpickle`. On the other hand, this solution may be useful if we actually need an "atomic" `_is_importable` check at some point in cloudpickle. I am pretty sure we do not need it, and that each time we are calling `_is_importable` right now in cloudpickle, we are actually checking for authorization to pickle by value, and not actual importability. Feel free to double check - if you confirm that, then I'm -0.5 on the solution because of the risk of work duplication.  If actual importability checks are needed, then this solution may achieve the best readability/efficiency tradeoff. 

An alternative fix is to rename `_is_importable` to `_should_pickle_by_value`, and:
- add a comment above the `lookup_module_and_qualname` call inside it precising that this call is a proxy to check importability of `obj`
- include a `is_registered_pickle_by_value` check inside `_should_pickle_by_value` that will use as input the module name returned as part of the result of the `_lookup_qualname_and_module` call made prior to this check.

Reply:
I can confirm that all five uses of `_is_importable` are contextually just checks to see if the pickle should be done by reference or value. Based on that, I'll rename `_is_importable` and update the logic as you suggest

Reply:
@pierreglaser - When implementing this, what is the "small additional adaptation for TypeVar pickling" you are referring to that I should add? Obviously something to get around the `cannot check importability of TypeVar instances` issue I'm getting? What should be the correct behaviour be? 

I've added  an `isinstance(obj, typing.TypeVar)` (upon which it does the lookup module) but I'm not familiar with this aspect and feel like this might be incorrect. What is different about TypeVars that gives rise to the different treatment?

Reply:
The adaptation I was referring to was that the logic effectively checking the import ability of a specific `TypeVar` instance is currently done outside of `_is_importable` which was fine back when the `is_registered_pickle_by_value` call was done as part of `_lookup_module_and_qualname`, but not anymore since we moved that call inside `_is_importable` (see `_typevar_reduce`, you'll understand).

As a result, `_is_importable` should be enhanced to support checking the importability of `TypeVar` instances, by simply porting the check inside `_typevar_reduce` inside `_is_importable`. Maybe there is something I am missing, but I don't think so.

Reply:
Im not sure I follow. The comments in `_typevar_reduce` indicate the the `__name__` is being explicitly passed in because there is no `__qualname__`, however checking for `__name__` when `__qualname__` is not found is already the default behaviour of the `_lookup_module_and_qualname`, so I don't understand why its separated out. 

Should I modify the `_typevar_reduce` to invoke `_should_pickle_by_reference` like `_function_reduce`?

Reply:
Mmh, yes sorry, I forgot that we actually need the `module` information inside `_typevar_reduce`. In that case, maybe just add a `is_registered_pickle_by_value` check inside the `_typevar_reduce`?